src/main/java/org/elasticsearch/indices/store/IndicesStore.java

```diff
@@ -318,38 +294,11 @@ public ClusterState execute(ClusterState currentState) throws Exception {
                         logger.trace("not deleting shard {}, the update task state version[{}] is not equal to cluster state before shard active api call [{}]", shardId, currentState.getVersion(), clusterState.getVersion());
                         return currentState;
                     }
-
```

Comment:
I wonder if we can remove more code here. We already have the indexUUID of the shard in question (it's part of the cluster state) and we check it on the remote nodes. if they acked they have the shard we can skip the cluster state checks and update on the state update task but rather just instruct IndicesService to verify this indexUUID is still current (or no other one exists) and delete the shard content if we can get the shard lock. This will things much simpler imho.